File: src/base-algorithms/dijkstra/typescript/dijkstra.ts
import MinHeap from "@/base-algorithms/utils/min_heap";

/**
 * Build an adjacency list for an undirected weighted graph.
 * Each `edges[i] = [ui, vi, wt]` adds a bidirectional edge of weight `wt`.
 */
function constructAdj(edges: number[][], n: number) {
  // adj[u] = list of [v, wt]
  const adj: [number, number][][] = Array.from({ length: n }, () => []);

  for (const edge of edges) {
    const [u, v, wt] = edge;
    adj[u].push([v, wt]);
    adj[v].push([u, wt]);
  }

  return adj;
}

/**
 * Compute shortest-path distances from `src` to every node using Dijkstra's algorithm.
 * @returns array where index `i` yields the minimum cost from `src` to `i`.
 */
export function dijkstra(n: number, edges: number[][], src: number) {
  // Create adjacency list
  const adj = constructAdj(edges, n);

  // Create a min heap to store <distance, node>
  const minHeap = new MinHeap();

  // Create an array for distances and initialize all distances as infinity
  const dist = Array(n).fill(Number.MAX_SAFE_INTEGER);

  // Push the source node with distance 0
  minHeap.push([0, src]);
  dist[src] = 0;

  // Pop the closest node from the PQ and relax its edges
  while (minHeap.size() > 0) {
    const item = minHeap.pop() as [number, number];

    // Traverse all adjacent of the node
    // item[1] is node label
    for (const [v, weight] of adj[item[1]]) {
      if (dist[v] > dist[item[1]] + weight) {
        dist[v] = dist[item[1]] + weight;
        minHeap.push([dist[v], v]);
      }
    }
  }

  return dist;
}

File: src/base-algorithms/dijkstra/cpp/src/dijkstra.cpp
#include "dijkstra.hpp"
#include <vector>
#include <queue>
#include <climits>
using namespace std;

// Function to construct adjacency 
vector<vector<vector<int>>> constructAdj(vector<vector<int>> 
                             &edges, int n) {
                                 
    // adj[u] = list of {v, wt}
    vector<vector<vector<int>>> adj(n); 

    for (const auto &edge : edges) {
        int u = edge[0];
        int v = edge[1];
        int wt = edge[2];
        adj[u].push_back({v, wt});
        adj[v].push_back({u, wt}); 
    }
    
    return adj;
}

// Returns shortest distances from src to all other vertices
vector<int> dijkstra(int n, vector<vector<int>> &edges, int src){
    
    // Create adjacency list
    vector<vector<vector<int>>> adj = constructAdj(edges, n);

    // Create a priority queue to store vertices that
    // are being preprocessed.
    priority_queue<vector<int>, vector<vector<int>>, 
                   greater<vector<int>>> pq;

    // Create a vector for distances and initialize all
    // distances as infinite
    vector<int> dist(n, INT_MAX);

    // Insert source itself in priority queue and initialize
    // its distance as 0.
    pq.push({0, src});
    dist[src] = 0;

    // Looping till priority queue becomes empty (or all
    // distances are not finalized) 
    while (!pq.empty()){
        
        // The first vertex in pair is the minimum distance
        // vertex, extract it from priority queue.
        int u = pq.top()[1];
        pq.pop();

        // Get all adjacent of u.
        for (auto x : adj[u]){
            
            // Get vertex label and weight of current
            // adjacent of u.
            int v = x[0];
            int weight = x[1];

            // If there is shorter path to v through u.
            if (dist[v] > dist[u] + weight)
            {
                // Updating distance of v
                dist[v] = dist[u] + weight;
                pq.push({dist[v], v});
            }
        }
    }

    return dist;
}
